### PR TITLE
add tests for sharpness, overprediction, and underprediction individu…

### PR DESCRIPTION
**Pull request by starkari**
…ally

issue #134

### CODE REVIEW

File: tests/testthat/test-score_forecasts.R [hunk 1]
```diff
@@ -391,6 +391,1366 @@ test_that("abs error is correct, point and median forecasts same", {
 
 })
 
+test_that("sharpness is correct, point forecast only, use_median_as_point FALSE", {
```
Comment (nickreich) [inline tests/testthat/test-score_forecasts.R:394]: as in other PR, suggest titling test "sharpness is NaN"...

File: tests/testthat/test-score_forecasts.R [hunk 2]
```diff
@@ -391,6 +391,1366 @@ test_that("abs error is correct, point and median forecasts same", {
...
+  )
+
+  actual <- score_forecasts(forecasts = test_forecasts, truth = test_truth, use_median_as_point = FALSE)
+
+  expected <- rep(NaN, length(point_forecast))
+
+  expect_equal(actual$sharpness, expected)
+})
+
+test_that("sharpness is correct, median only, use_median_as_point TRUE", {
```
Comment (nickreich) [inline tests/testthat/test-score_forecasts.R:443]: "sharpness is zero"

File: tests/testthat/test-score_forecasts.R [hunk 3]
```diff
@@ -391,6 +391,1366 @@ test_that("abs error is correct, point and median forecasts same", {
...
+  )
+  actual <- score_forecasts(
+    forecasts = test_forecasts, truth = test_truth,
+    use_median_as_point = TRUE
+  )
+  expected <- rep(0, length(y))
+  expect_equal(actual$sharpness, expected)
+})
+
+test_that("sharpness is correct, median only, use_median_as_point FALSE", {
```
Comment (nickreich) [inline tests/testthat/test-score_forecasts.R:496]: "sharpness is zero"

File: tests/testthat/test-score_forecasts.R [hunk 4]
```diff
@@ -391,6 +391,1366 @@ test_that("abs error is correct, point and median forecasts same", {
...
+    stringsAsFactors = FALSE
+  )
+
+  expect_error(actual <- score_forecasts(
+    forecasts = test_forecasts, truth = test_truth,
+    use_median_as_point = FALSE
+  ))
+})
+
+test_that("sharpness is correct, 1 interval only, use_median_as_point TRUE", {
```
Comment (nickreich) [inline tests/testthat/test-score_forecasts.R:552]: This test seems to have two intervals in the data, although maybe only one is tested? seems like there should be alignment between the data passed and the test.

File: tests/testthat/test-score_forecasts.R [hunk 5]
```diff
@@ -391,6 +391,1366 @@ test_that("abs error is correct, point and median forecasts same", {
...
+    temporal_resolution = rep("wk", n_forecasts),
+    target_variable = forecast_target_variables,
+    target_end_date = forecast_target_end_dates,
+    type = rep("quantile", n_forecasts),
+    quantile = forecast_quantile_probs,
+    value = forecast_quantiles,
+    stringsAsFactors = FALSE
+  )
+
+  expect_error(actual <- score_forecasts(
```
Comment (nickreich) [inline tests/testthat/test-score_forecasts.R:654]: why are we expecting error here?

File: tests/testthat/test-score_forecasts.R [hunk 6]
```diff
@@ -391,6 +391,1366 @@ test_that("abs error is correct, point and median forecasts same", {
...
+  alpha2 <- 0.5
+  expected <- (1 / 2.5) * (
+    abs(forecast_quantiles_matrix[, 5] - forecast_quantiles_matrix[, 1]) * (alpha1 / 2) +
+      abs(forecast_quantiles_matrix[, 4] - forecast_quantiles_matrix[, 2]) * (alpha2 / 2)
+  )
+
+  expect_equal(actual$sharpness, expected)
+})
+
+test_that("overprediction is correct, point forecast only, use_median_as_point FALSE", {
```
Comment (nickreich) [inline tests/testthat/test-score_forecasts.R:845]: overprediction is NaN

File: tests/testthat/test-score_forecasts.R [hunk 7]
```diff
@@ -391,6 +391,1366 @@ test_that("abs error is correct, point and median forecasts same", {
...
+  )
+  actual <- score_forecasts(
+    forecasts = test_forecasts, truth = test_truth,
+    use_median_as_point = TRUE
+  )
+  expected <- (forecast_quantiles_matrix[, 1] - y) * (y < forecast_quantiles_matrix[, 1])
+  expect_equal(actual$overprediction, expected)
+})
+
+test_that("overprediction is correct, median only, use_median_as_point FALSE", {
```
Comment (nickreich) [inline tests/testthat/test-score_forecasts.R:947]: not clear to me why we should expect an error here.

File: tests/testthat/test-score_forecasts.R [hunk 8]
```diff
@@ -391,6 +391,1366 @@ test_that("abs error is correct, point and median forecasts same", {
...
+    type = rep("quantile", n_forecasts),
+    quantile = forecast_quantile_probs,
+    value = forecast_quantiles,
+    stringsAsFactors = FALSE
+  )
+  actual <- score_forecasts(
+    forecasts = test_forecasts, truth = test_truth,
+    use_median_as_point = TRUE
+  )
+  alpha1 <- 0.2
```
Comment (nickreich) [inline tests/testthat/test-score_forecasts.R:1052]: not important, but alpha1 is not used in the calculations.

File: tests/testthat/test-score_forecasts.R [hunk 9]
```diff
@@ -391,6 +391,1366 @@ test_that("abs error is correct, point and median forecasts same", {
...
+  expected <- (forecast_quantiles_matrix[, 1] - y) * (y < forecast_quantiles_matrix[, 1])
+  expect_equal(actual$overprediction, expected)
+})
+
+test_that("overprediction is correct, 2 intervals and median no point, use_median_as_point FALSE", {
```
Comment (nickreich) [inline tests/testthat/test-score_forecasts.R:1057]: why error here?

File: tests/testthat/test-score_forecasts.R [hunk 10]
```diff
@@ -391,6 +391,1366 @@ test_that("abs error is correct, point and median forecasts same", {
...
+    type = rep("quantile", n_forecasts),
+    quantile = forecast_quantile_probs,
+    value = forecast_quantiles,
+    stringsAsFactors = FALSE
+  )
+  actual <- score_forecasts(
+    forecasts = test_forecasts, truth = test_truth,
+    use_median_as_point = TRUE
+  )
+  alpha1 <- 0.2
```
Comment (nickreich) [inline tests/testthat/test-score_forecasts.R:1158]: alpha1 alpha2 not used

File: tests/testthat/test-score_forecasts.R [hunk 11]
```diff
@@ -391,6 +391,1366 @@ test_that("abs error is correct, point and median forecasts same", {
...
+  expected <- (1 / 2.5) * (
+    (1 / 2) * (forecast_quantiles_matrix[, 3] - y) * (y < forecast_quantiles_matrix[, 3]) +
+      (forecast_quantiles_matrix[, 1] - y) * (y < forecast_quantiles_matrix[, 1]) +
+      (forecast_quantiles_matrix[, 2] - y) * (y < forecast_quantiles_matrix[, 2])
+  )
+
+  expect_equal(actual$overprediction, expected)
+})
+
+test_that("underprediction is correct, point forecast only, use_median_as_point FALSE", {
```
Comment (nickreich) [inline tests/testthat/test-score_forecasts.R:1299]: is NaN

File: tests/testthat/test-score_forecasts.R [hunk 12]
```diff
@@ -391,6 +391,1366 @@ test_that("abs error is correct, point and median forecasts same", {
...
+  )
+  actual <- score_forecasts(
+    forecasts = test_forecasts, truth = test_truth,
+    use_median_as_point = TRUE
+  )
+  expected <- (y - forecast_quantiles_matrix[, 1]) * (y > forecast_quantiles_matrix[, 1])
+  expect_equal(actual$underprediction, expected)
+})
+
+test_that("underprediction is correct, median only, use_median_as_point FALSE", {
```
Comment (nickreich) [inline tests/testthat/test-score_forecasts.R:1401]: why error?

**Review by nickreich — CHANGES_REQUESTED** [review]
mostly minor changes, but curious about why we should expect error in some cases when use_median_as_point=FALSE.

**Comment by starkari** [comment]
> mostly minor changes, but curious about why we should expect error in some cases when use_median_as_point=FALSE.

it errors because it is expecting a point forecast to be passed in but there is no point forecast

File: tests/testthat/test-score_forecasts.R (hunk 4 above)
Comment (starkari) [reply tests/testthat/test-score_forecasts.R:552]: The finalized data is only 1 interval alpha=0.2 so quantiles 0.1 and 0.9. I tried to remain consistent with the set up of wis.

**Review by nickreich — APPROVED** [review]

**Comment by nickreich** [comment]
dang, now there is a conflict here. @starkari can you resolve and merge this branch?

**Pull request merged**